cpp/src/barretenberg/benchmark/honk_bench/benchmark_utilities.hpp

```diff
@@ -0,0 +1,224 @@
+    using fr = typename curve::fr;
+    using fq = typename curve::fq;
+    using g1 = typename curve::g1;
+
+    std::string message_string = "Instructions unclear, ask again later.";
+
+    crypto::ecdsa::key_pair<fr, g1> account;
+    account.private_key = curve::fr::random_element();
+    account.public_key = curve::g1::one * account.private_key;
+
+    crypto::ecdsa::signature signature =
+        crypto::ecdsa::construct_signature<Sha256Hasher, fq, fr, g1>(message_string, account);
+
+    bool first_result =
+        crypto::ecdsa::verify_signature<Sha256Hasher, fq, fr, g1>(message_string, account.public_key, signature);
+
+    std::vector<uint8_t> rr(signature.r.begin(), signature.r.end());
+    std::vector<uint8_t> ss(signature.s.begin(), signature.s.end());
+    uint8_t vv = signature.v;
+
+    typename curve::g1_bigfr_ct public_key = curve::g1_bigfr_ct::from_witness(&composer, account.public_key);
+
+    proof_system::plonk::stdlib::ecdsa::signature<Composer> sig{ typename curve::byte_array_ct(&composer, rr),
+                                                                 typename curve::byte_array_ct(&composer, ss),
+                                                                 proof_system::plonk::stdlib::uint8<Composer>(&composer,
+                                                                                                              vv) };
+
+    typename curve::byte_array_ct message(&composer, message_string);
+
+    for (size_t i = 0; i < num_iterations; i++) {
+        proof_system::plonk::stdlib::ecdsa::verify_signature<Composer,
+                                                             curve,
+                                                             typename curve::fq_ct,
+                                                             typename curve::bigfr_ct,
+                                                             typename curve::g1_bigfr_ct>(message, public_key, sig);
+    }
+}
+
+/**
+ * @brief Generate test circuit with specified number of merkle membership checks
+ *
+ * @param composer
+ * @param num_iterations
+ * @todo (luke): should we consider deeper tree? non-zero leaf values? variable index?
+ */
+template <typename Composer> void generate_merkle_membership_test_circuit(Composer& composer, size_t num_iterations)
+{
+    using namespace proof_system::plonk::stdlib;
+    using field_ct = field_t<Composer>;
+    using witness_ct = witness_t<Composer>;
+    using witness_ct = witness_t<Composer>;
+    using MemStore = merkle_tree::MemoryStore;
+    using MerkleTree_ct = merkle_tree::MerkleTree<MemStore>;
+
+    MemStore store;
+    auto db = MerkleTree_ct(store, 3);
+
+    // Check that the leaf at index 0 has value 0.
+    auto zero = field_ct(witness_ct(&composer, fr::zero())).decompose_into_bits();
+    field_ct root = witness_ct(&composer, db.root());
+
+    for (size_t i = 0; i < num_iterations; i++) {
+        merkle_tree::check_membership(
```

Comment:
Here, too. Ideally for several iterations of benchmarks it shouldn't be absolutely the same action. I assume we'll add filtering at some point and a repeat action on the same witness indices would not be lead to additional gates

Reply:
Updated to check a different leaf each time